<a href="https://colab.research.google.com/github/MarkowitzMx/Programacion-para-analitica-descriptiva-y-predictiva-2026/blob/main/Copy_of_Actividad_Calificada_07_Datos_Duplicados_274690.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **Actividad calificada: Datos duplicados**
### Detección, cuantificación y tratamiento de duplicados en un dataset real

---

| | |
|---|---|
| **Materia** | Programación para Analítica Descriptiva y Predictiva |
| **Programa** | Maestría en Inteligencia Artificial y Analítica de Datos (MIAAD) — UACJ |
| **Unidad** | 02 Análisis Descriptivo de los Datos |
| **Tema** | Limpieza y transformación de datos: duplicados |
| **Modalidad** | Individual |
| **Valor** | 100 puntos |
| **Dataset** | Hotel Booking Demand (Antonio, Almeida y Nunes, 2019) |
| **Fecha de entrega** | 19 Septiembre 2026 |

---

## Datos del estudiante

| | |
|---|---|
| **Nombre completo** Cristóbal Lemus Rendón| |   
| **Matrícula**                       274690| |  

## Instrucciones

1. Trabaja en Google Colab. Guarda una copia en tu Drive (**Archivo → Guardar una copia en Drive**) antes de empezar.
2. Completa tus datos en la tabla anterior.
3. Resuelve las actividades **en orden**: algunas usan variables creadas en actividades anteriores.
4. Escribe tu código en las celdas marcadas con `# Escribe tu código aquí`. Puedes agregar celdas si lo necesitas.
5. **Comenta tu código**: cada bloque debe explicar qué hace y por qué. Los comentarios forman parte de la calificación de cada actividad.
6. Responde las preguntas en las celdas **Respuesta**, con tus propias palabras y apoyándote en los resultados que obtuviste (cita cifras).
7. Usa los métodos vistos en el notebook de la sesión: `duplicated()`, `drop_duplicates()`, `value_counts()`, `sort_values()`, `groupby()`, `pd.concat()`, máscaras booleanas, etc.
8. Antes de entregar, ejecuta todo el notebook (**Entorno de ejecución → Reiniciar y ejecutar todo**) y verifica que no haya errores.
9. **Entrega en Moodle** un archivo de texto `Actividad_Duplicados_<Matricula>.txt` que contenga el **enlace** a tu notebook (Google Colab o GitHub), con las salidas visibles.
   - Colab: **Compartir → Acceso general → Cualquier persona con el enlace (Lector)**.
   - GitHub: el repositorio debe ser **público**.
   - Opcional: puedes adjuntar también el archivo `.ipynb`.

## Contexto

El dataset **Hotel Booking Demand** contiene 119,390 reservas de dos hoteles de Portugal (un hotel urbano, *City Hotel*, y un hotel vacacional, *Resort Hotel*) con llegada entre julio de 2015 y agosto de 2017. Cada fila es una reserva y hay 32 columnas. Algunas relevantes para esta actividad:

| Columna | Descripción |
|---|---|
| `hotel` | `City Hotel` o `Resort Hotel` |
| `is_canceled` | 1 si la reserva se canceló, 0 si no |
| `lead_time` | Días entre la reserva y la fecha de llegada |
| `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month` | Fecha de llegada |
| `adults`, `children`, `babies` | Número de huéspedes |
| `country` | País de origen (código ISO de 3 letras) |
| `market_segment` | Segmento de mercado (`Online TA`, `Groups`, `Direct`, …) |
| `deposit_type` | Tipo de depósito (`No Deposit`, `Non Refund`, `Refundable`) |
| `agent`, `company` | Identificador de la agencia o empresa que hizo la reserva (con faltantes) |
| `customer_type` | Tipo de cliente (`Transient`, `Group`, …) |
| `adr` | Tarifa diaria promedio (*average daily rate*) |
| `reservation_status` | Estado final: `Check-Out`, `Canceled`, `No-Show` |
| `reservation_status_date` | Fecha del último cambio de estado |

> **Importante**: el dataset **no incluye un identificador de reserva**. Tenlo presente al interpretar los duplicados.

## Rúbrica

| Actividad | Puntos | Excelente | Bueno | Insuficiente | No entregado / Incorrecto |
|---|---|---|---|---|---|
| **1. Diagnóstico inicial** | 20 | 18–20: conteos y porcentajes correctos con los tres valores de `keep`; duplicados por hotel calculados; código comentado; conclusión con cifras. | 14–17: resultados correctos con omisiones menores en comentarios o en la conclusión. | 1–13: conteos incompletos o con errores; conclusión sin sustento. | 0: no se entregó o el código no produce resultados. |
| **2. Inspección y criterio** | 25 | 23–25: grupos inspeccionados y ordenados; tamaños de grupo analizados; comparación duplicados vs. total; argumento sólido sobre si son errores, con evidencia. | 18–22: análisis correcto pero con argumento parcial o poca evidencia. | 1–17: inspección superficial; argumento sin datos. | 0: no se entregó o es incorrecto. |
| **3. Impacto en el análisis** | 25 | 23–25: métricas por hotel antes y después con `groupby()`; tabla comparativa con `pd.concat()`; interpretación del efecto en el análisis y en un modelo predictivo. | 18–22: métricas correctas; tabla o interpretación incompletas. | 1–17: métricas con errores o sin comparación. | 0: no se entregó o es incorrecto. |
| **4. Tratamiento y verificación** | 20 | 18–20: limpieza verificada; duplicados parciales detectados e interpretados; registro de decisión completo. | 14–17: limpieza correcta; análisis parcial o registro incompleto. | 1–13: limpieza sin verificación o análisis parcial erróneo. | 0: no se entregó o es incorrecto. |
| **Reflexión final** | 10 | 9–10: respuestas fundamentadas en los resultados propios. | 7–8: respuestas correctas pero generales. | 1–6: respuestas vagas o sin relación con los resultados. | 0: no se entregó. |
| **Total** | **100** | | | | |

**Notas**:

- Las respuestas escritas de cada actividad (1.4, 2.4, 3.4, 4.2 y 4.3) y los comentarios del código cuentan para la calificación **de esa actividad**, no solo para la reflexión final.
- Entrega tardía: se descuenta **5 % por día natural** de retraso, hasta 7 días; después de ese plazo la calificación es 0. El descuento se aplica sobre la calificación obtenida.

## Carga de datos

Ejecuta la siguiente celda. No requiere modificaciones.

In [12]:
import pandas as pd

# Carga del dataset Hotel Booking Demand desde GitHub (TidyTuesday, 2020-02-11)
url = 'https://raw.githubusercontent.com/rfordatascience/tidytuesday/master/data/2020/2020-02-11/hotels.csv'
hoteles = pd.read_csv(url)

# Dimensiones y primeras filas
print('Dimensiones:', hoteles.shape)
hoteles.head()

Dimensiones: (119390, 32)


,hotel,is_canceled,lead_time,arrival_date_year,arrival_date_month,arrival_date_week_number,arrival_date_day_of_month,stays_in_weekend_nights,stays_in_week_nights,adults,...,deposit_type,agent,company,days_in_waiting_list,customer_type,adr,required_car_parking_spaces,total_of_special_requests,reservation_status,reservation_status_date
0,Resort Hotel,0,342,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
1,Resort Hotel,0,737,2015,July,27,1,0,0,2,...,No Deposit,NaN,NaN,0,Transient,0.0,0,0,Check-Out,2015-07-01
2,Resort Hotel,0,7,2015,July,27,1,0,1,1,...,No Deposit,NaN,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
3,Resort Hotel,0,13,2015,July,27,1,0,1,1,...,No Deposit,304.0,NaN,0,Transient,75.0,0,0,Check-Out,2015-07-02
4,Resort Hotel,0,14,2015,July,27,1,0,2,2,...,No Deposit,240.0,NaN,0,Transient,98.0,0,1,Check-Out,2015-07-03


---
## Actividad 1. Diagnóstico inicial (20 puntos)

### 1.1 Conteo de duplicados exactos (5 pts)

Calcula el número de filas marcadas como duplicadas en `hoteles` considerando **todas** las columnas, con `keep='first'`, `keep='last'` y `keep=False`. Imprime los tres resultados con un texto que explique qué significa cada uno.

In [13]:
# Contamos las filas duplicadas considerando TODAS las columnas del dataset.
# keep='first' -> marca como duplicada cada copia EXCEPTO la primera aparicion de cada combinacion.
#                 el resultado es el numero de filas "sobrantes" si conservaramos solo la primera copia.
# keep='last'  -> marca como duplicada cada copia EXCEPTO la ultima aparicion.
#                 por eso el conteo es el mismo numero que con 'first' (es la misma cantidad de "sobrantes",
#                 solo cambia CUAL copia se deja como la "original").
# keep=False   -> marca como duplicada CUALQUIER fila que tenga al menos una copia identica en otro lugar,
#                 es decir, TODAS las filas que participan en un grupo repetido (incluida la primera y la ultima).

dup_first = hoteles.duplicated(keep='first').sum()
dup_last = hoteles.duplicated(keep='last').sum()
dup_false = hoteles.duplicated(keep=False).sum()

print(f"keep='first' -> {dup_first} filas duplicadas (filas 'sobrantes' si nos quedamos con la primera copia)")
print(f"keep='last'  -> {dup_last} filas duplicadas (filas 'sobrantes' si nos quedamos con la ultima copia)")
print(f"keep=False   -> {dup_false} filas duplicadas (TODAS las filas involucradas en algun grupo repetido)")


keep='first' -> 31994 filas duplicadas (filas 'sobrantes' si nos quedamos con la primera copia)
keep='last'  -> 31994 filas duplicadas (filas 'sobrantes' si nos quedamos con la ultima copia)
keep=False   -> 40165 filas duplicadas (TODAS las filas involucradas en algun grupo repetido)


### 1.2 Magnitud del problema (5 pts)

Con los resultados de 1.1, calcula e imprime:

- El **porcentaje** de filas sobrantes respecto al total.
- El número de **grupos repetidos** (combinaciones de valores que aparecen más de una vez).

In [14]:
# Porcentaje de filas "sobrantes" (las que se eliminarian si depuraramos el dataset)
# respecto al total de filas. Usamos el resultado de keep='first' de la actividad 1.1.
pct_sobrantes = dup_first / len(hoteles) * 100
print(f"Porcentaje de filas sobrantes respecto al total: {pct_sobrantes:.2f}%")

# Numero de grupos repetidos: combinaciones de valores que aparecen mas de una vez.
# Tomamos todas las filas que participan en un duplicado (keep=False) y eliminamos las copias
# dentro de ese subconjunto (drop_duplicates) para quedarnos con UNA fila por cada combinacion distinta.
# El numero de filas resultante es el numero de grupos repetidos.
filas_en_grupos = hoteles[hoteles.duplicated(keep=False)]
n_grupos_repetidos = filas_en_grupos.drop_duplicates().shape[0]
print(f"Numero de grupos repetidos (combinaciones distintas que se repiten): {n_grupos_repetidos}")


Porcentaje de filas sobrantes respecto al total: 26.80%
Numero de grupos repetidos (combinaciones distintas que se repiten): 8171


### 1.3 Duplicados por hotel (5 pts)

Determina **cuántas** filas sobrantes (`keep='first'`) hay en cada hotel y qué **porcentaje** representan respecto al total de reservas de ese hotel.

Sugerencia: crea una copia de `hoteles`, agrega una columna booleana `es_duplicado` con el resultado de `duplicated()` y agrupa por `hotel`.

In [15]:
# Copiamos el DataFrame para no modificar el original y agregamos una columna booleana
# que indica si la fila es un duplicado, usando keep='first' (asi cada grupo aporta el numero
# de copias "sobrantes" respecto a su primera aparicion).
hoteles_marcado = hoteles.copy()
hoteles_marcado['es_duplicado'] = hoteles_marcado.duplicated(keep='first')

# Agrupamos por hotel y calculamos: cuantas filas sobrantes hay (suma de la columna booleana,
# porque True se comporta como 1) y cuantas reservas totales tiene cada hotel (count).
duplicados_por_hotel = hoteles_marcado.groupby('hotel')['es_duplicado'].agg(
    filas_duplicadas='sum',
    total_reservas='count'
)

# El porcentaje se calcula respecto al total de CADA hotel (no respecto al total general),
# para poder comparar la incidencia del problema entre los dos hoteles de forma justa.
duplicados_por_hotel['porcentaje'] = (
    duplicados_por_hotel['filas_duplicadas'] / duplicados_por_hotel['total_reservas'] * 100
).round(2)

print(duplicados_por_hotel)


              filas_duplicadas  total_reservas  porcentaje
hotel                                                     
City Hotel               25902           79330       32.65
Resort Hotel              6092           40060       15.21


### 1.4 Conclusión (5 pts)

¿Qué tan grave es el problema de duplicados en este dataset? ¿Afecta por igual a ambos hoteles? Responde con cifras.

**Respuesta 1.4:**

El problema de duplicados es considerable: 31,994 filas (26.80% del total de 119,390 reservas) son copias exactas de otra fila ya registrada (`keep='first'`), y en total 40,165 filas (33.65%) participan en algun grupo repetido (`keep=False`), agrupadas en 8,171 combinaciones distintas de valores que aparecen mas de una vez.

El problema **no afecta por igual a ambos hoteles**. En el City Hotel, 25,902 de 79,330 reservas son duplicados sobrantes (32.65%), mientras que en el Resort Hotel son 6,092 de 40,060 (15.21%). Es decir, la proporcion de duplicados en el City Hotel es mas del doble que en el Resort Hotel, lo que sugiere que el problema podria estar relacionado con el canal o proceso de captura de reservas de ese hotel en particular (por ejemplo, mas reservas por agencias/grupos, como se explora en la Actividad 2).


---
## Actividad 2. Inspección y criterio de duplicado (25 puntos)

### 2.1 Ver los grupos repetidos (6 pts)

Filtra las filas involucradas en duplicados (`keep=False`), **ordénalas** con `sort_values()` para que las copias queden juntas y muestra las primeras 12 filas con estas columnas:

`hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adults`, `country`, `market_segment`, `deposit_type`, `adr`, `reservation_status`.

Elige tú las columnas de ordenamiento y explica tu elección en un comentario.

In [16]:
# Seleccionamos todas las filas que participan en algun duplicado (keep=False incluye
# la primera, la ultima y cualquier copia intermedia de cada grupo repetido).
columnas_vista = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                   'arrival_date_day_of_month', 'adults', 'country', 'market_segment',
                   'deposit_type', 'adr', 'reservation_status']

filas_duplicadas = hoteles[hoteles.duplicated(keep=False)]

# Ordenamos por hotel y por la fecha de llegada (anio, mes, dia) y luego por lead_time y adr.
# Elegimos estas columnas porque son las que en teoria deberian coincidir entre copias de una
# MISMA reserva (mismo hotel, misma fecha de llegada). Ordenar por ellas hace que las filas
# identicas queden fisicamente juntas al imprimir la tabla, facilitando la inspeccion visual
# de cada grupo repetido.
filas_duplicadas_ordenadas = filas_duplicadas.sort_values(
    by=['hotel', 'arrival_date_year', 'arrival_date_month', 'arrival_date_day_of_month', 'lead_time', 'adr']
)

filas_duplicadas_ordenadas[columnas_vista].head(12)


,hotel,lead_time,arrival_date_year,arrival_date_month,arrival_date_day_of_month,adults,country,market_segment,deposit_type,adr,reservation_status
40574,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40580,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40588,City Hotel,2,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40562,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40585,City Hotel,117,2015,August,1,1,ITA,Offline TA/TO,No Deposit,72.5,Check-Out
40563,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40564,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled
40566,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40567,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Check-Out
40568,City Hotel,117,2015,August,1,2,ITA,Offline TA/TO,No Deposit,94.5,Canceled


### 2.2 Tamaño de los grupos (6 pts)

Usa `value_counts()` sobre el `DataFrame` completo para contar cuántas veces aparece cada combinación de filas. Considera que las columnas `agent` y `company` tienen faltantes (revisa el parámetro `dropna`).

Con ese resultado:

1. Muestra **cuántos grupos** hay de cada tamaño (2 copias, 3 copias, …). Basta con los 10 tamaños más frecuentes.
2. Indica el **tamaño del grupo más grande** y muestra su combinación de valores.

In [17]:
# value_counts() sobre el DataFrame completo cuenta cuantas veces se repite cada
# combinacion unica de valores (fila completa). Usamos dropna=False porque las columnas
# 'agent' y 'company' tienen muchos NaN: si no lo indicamos, pandas podria tratar de forma
# especial esas combinaciones con NaN y no contarlas como iguales entre si por defecto en
# versiones antiguas; con dropna=False forzamos a que los NaN se consideren un valor mas
# al comparar, igual que hace duplicated().
conteo_combinaciones = hoteles.value_counts(dropna=False)

# 1. Cuantos grupos hay de cada tamano (2 copias, 3 copias, etc.).
# Tomamos la distribucion de los propios conteos: cuantas combinaciones (grupos) tienen
# tamano 1, cuantas tienen tamano 2, etc. Mostramos los 10 tamanos mas frecuentes.
tamanos_de_grupo = conteo_combinaciones.value_counts().sort_values(ascending=False)
print("Cantidad de grupos segun su tamano (10 mas frecuentes):")
print(tamanos_de_grupo.head(10))

# 2. Tamano del grupo mas grande y su combinacion de valores.
tamano_maximo = conteo_combinaciones.max()
print(f"\nEl grupo mas grande tiene {tamano_maximo} copias identicas.")
print("\nCombinacion de valores de ese grupo:")
print(conteo_combinaciones[conteo_combinaciones == tamano_maximo])


Cantidad de grupos segun su tamano (10 mas frecuentes):
count
1     79225
2      5034
3       969
4       497
5       275
6       171
10      123
7       115
8       103
20       97
Name: count, dtype: int64

El grupo mas grande tiene 180 copias identicas.

Combinacion de valores de ese grupo:
hotel       is_canceled  lead_time  arrival_date_year  arrival_date_month  arrival_date_week_number  arrival_date_day_of_month  stays_in_weekend_nights  stays_in_week_nights  adults  children  babies  meal  country  market_segment  distribution_channel  is_repeated_guest  previous_cancellations  previous_bookings_not_canceled  reserved_room_type  assigned_room_type  booking_changes  deposit_type  agent  company  days_in_waiting_list  customer_type  adr    required_car_parking_spaces  total_of_special_requests  reservation_status  reservation_status_date
City Hotel  1            277        2016               November            46                        7                          1                

### 2.3 ¿En qué se distinguen las filas duplicadas? (6 pts)

Compara las filas involucradas en duplicados (`keep=False`) con el dataset completo en:

- Distribución de `market_segment` (proporciones).
- Distribución de `deposit_type` (proporciones).
- Tasa de cancelación (`is_canceled`).

In [18]:
# Comparamos la distribucion de algunas variables clave entre:
#  a) el dataset completo (referencia)
#  b) el subconjunto de filas involucradas en duplicados (keep=False)

filas_duplicadas = hoteles[hoteles.duplicated(keep=False)]

# Proporciones de market_segment: normalize=True convierte los conteos en proporciones (0 a 1)
# para poder comparar directamente aunque el numero de filas de cada grupo sea distinto.
comparacion_market_segment = pd.concat([
    hoteles['market_segment'].value_counts(normalize=True).rename('dataset_completo'),
    filas_duplicadas['market_segment'].value_counts(normalize=True).rename('duplicados')
], axis=1).round(3)
print("=== Proporcion por market_segment ===")
print(comparacion_market_segment)

# Proporciones de deposit_type
comparacion_deposit_type = pd.concat([
    hoteles['deposit_type'].value_counts(normalize=True).rename('dataset_completo'),
    filas_duplicadas['deposit_type'].value_counts(normalize=True).rename('duplicados')
], axis=1).round(3)
print("\n=== Proporcion por deposit_type ===")
print(comparacion_deposit_type)

# Tasa de cancelacion: como is_canceled es 0/1, la media es directamente la proporcion de
# reservas canceladas.
print("\n=== Tasa de cancelacion (is_canceled) ===")
print(f"Dataset completo: {hoteles['is_canceled'].mean():.4f}")
print(f"Filas duplicadas: {filas_duplicadas['is_canceled'].mean():.4f}")


=== Proporcion por market_segment ===
                dataset_completo  duplicados
market_segment                              
Online TA                  0.473       0.206
Offline TA/TO              0.203       0.304
Groups                     0.166       0.416
Direct                     0.106       0.035
Corporate                  0.044       0.037
Complementary              0.006       0.002
Aviation                   0.002       0.000
Undefined                  0.000         NaN

=== Proporcion por deposit_type ===
              dataset_completo  duplicados
deposit_type                              
No Deposit               0.876       0.642
Non Refund               0.122       0.356
Refundable               0.001       0.002

=== Tasa de cancelacion (is_canceled) ===
Dataset completo: 0.3704
Filas duplicadas: 0.5839


### 2.4 Criterio (7 pts)

El dataset no tiene identificador de reserva. Con base en la evidencia de 2.1, 2.2 y 2.3, responde:

a) ¿Las filas duplicadas son **errores de registro** o podrían ser **reservas distintas** que coinciden en todos sus valores? Argumenta con datos.

b) Las columnas `agent` y `company` tienen muchos faltantes. ¿Cómo trata `duplicated()` esos `NaN` y qué implicación tiene para este análisis?

**Respuesta 2.4:**

**a)** La evidencia apunta mas a **errores de registro / reservas re-enviadas al sistema** que a reservas genuinamente distintas que coinciden por azar en todos sus valores. Tres hallazgos lo sustentan:

- En 2.2, existen grupos de hasta **180 filas identicas** (mismo hotel, misma fecha de llegada, mismo adr, mismo `reservation_status_date`, etc.). Es estadisticamente muy poco probable que 180 clientes distintos coincidan exactamente en las 32 variables del dataset (incluyendo columnas continuas como `adr`), lo que sugiere una reserva unica registrada muchas veces (por ejemplo, por reintentos del sistema o carga duplicada de datos).
- En 2.3, las filas duplicadas estan **sobre-representadas en `market_segment = 'Groups'`** (41.6% de los duplicados contra solo 16.6% en el dataset completo) y en `deposit_type = 'Non Refund'` (35.6% contra 12.2% en general). Esto es consistente con reservas grupales o con deposito no reembolsable, que suelen gestionarse por bloques y son mas susceptibles a duplicarse en la carga de datos.
- La tasa de cancelacion de las filas duplicadas (58.39%) es notablemente mayor que la del dataset completo (37.04%), lo que sugiere que el sistema pudo haber vuelto a registrar la reserva al momento de cancelarla o reprocesarla.

Sin un identificador de reserva no podemos probarlo con certeza absoluta, pero el patron (grupos muy grandes, concentracion en Groups/Non Refund, mayor cancelacion) es mas compatible con duplicacion tecnica que con coincidencia genuina entre reservas distintas.

**b)** `duplicated()` (y `drop_duplicates()`) tratan **`NaN == NaN` como verdadero** al comparar filas, es decir, dos valores faltantes en `agent` o `company` se consideran iguales entre si para efectos de determinar si una fila es duplicada. La implicacion es que, si dos reservas distintas simplemente no tienen agencia ni empresa registrada (`NaN` en ambas columnas) pero coinciden en el resto de las variables, `duplicated()` las marcara como duplicadas aunque en realidad podrian ser reservas diferentes cuyo unico "parecido" es la falta de ese dato. Esto puede **sobreestimar** el numero de duplicados, especialmente porque `agent` y `company` tienen muchos faltantes en este dataset.


---
## Actividad 3. Impacto en el análisis (25 puntos)

### 3.1 Versión sin duplicados (5 pts)

Crea `hoteles_limpio` eliminando los duplicados exactos, conservando la primera aparición y **renumerando el índice**. Imprime el número de filas antes y después y cuántas se eliminaron.

In [19]:
# Eliminamos los duplicados exactos conservando la primera aparicion de cada combinacion
# (keep='first' es el comportamiento por defecto de drop_duplicates, lo dejamos explicito
# por claridad). reset_index(drop=True) renumera el indice de 0 a n-1 para que quede
# consecutivo despues de haber quitado filas.
filas_antes = len(hoteles)

hoteles_limpio = hoteles.drop_duplicates(keep='first').reset_index(drop=True)

filas_despues = len(hoteles_limpio)
filas_eliminadas = filas_antes - filas_despues

print(f"Filas antes de eliminar duplicados: {filas_antes}")
print(f"Filas despues de eliminar duplicados: {filas_despues}")
print(f"Filas eliminadas: {filas_eliminadas}")


Filas antes de eliminar duplicados: 119390
Filas despues de eliminar duplicados: 87396
Filas eliminadas: 31994


### 3.2 Métricas por hotel (8 pts)

Para `hoteles` y para `hoteles_limpio`, calcula por hotel con `groupby()`:

- Número de reservas (`.size()`).
- Tasa de cancelación (media de `is_canceled`).
- Tarifa diaria promedio (media de `adr`).

Nota: para calcular la media de varias columnas a la vez, selecciona una lista de columnas: `df.groupby('hotel')[['col1', 'col2']].mean()`.

Guarda las medias en dos `DataFrame`, `resumen_original` y `resumen_limpio`, con `hotel` como columna (usa `reset_index()`).

In [20]:
# Calculamos, para cada version del dataset, tres metricas por hotel:
# tamano del grupo (numero de reservas), tasa de cancelacion promedio y adr promedio.
# Usamos .agg() con un diccionario de nombre->(columna, funcion) para nombrar las columnas
# resultantes directamente y evitar un MultiIndex en las columnas.

resumen_original = hoteles.groupby('hotel').agg(
    n_reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    adr_promedio=('adr', 'mean')
).reset_index()

resumen_limpio = hoteles_limpio.groupby('hotel').agg(
    n_reservas=('hotel', 'size'),
    tasa_cancelacion=('is_canceled', 'mean'),
    adr_promedio=('adr', 'mean')
).reset_index()

print("Resumen con datos originales:")
print(resumen_original)
print("\nResumen sin duplicados:")
print(resumen_limpio)


Resumen con datos originales:
          hotel  n_reservas  tasa_cancelacion  adr_promedio
0    City Hotel       79330          0.417270    105.304465
1  Resort Hotel       40060          0.277634     94.952930

Resumen sin duplicados:
          hotel  n_reservas  tasa_cancelacion  adr_promedio
0    City Hotel       53428          0.300386    110.985944
1  Resort Hotel       33968          0.234809     99.025346


### 3.3 Tabla comparativa (6 pts)

Agrega a cada resumen una columna `version` (`'original'` y `'sin duplicados'`) y únelos con `pd.concat()` en un `DataFrame` llamado `comparacion`, con índice renumerado y ordenado por `hotel`, de modo que las dos versiones de cada hotel queden juntas.

In [21]:
# Agregamos una columna 'version' a cada resumen para poder distinguir el origen de cada
# fila una vez que los unamos en un solo DataFrame.
resumen_original_v = resumen_original.copy()
resumen_original_v['version'] = 'original'

resumen_limpio_v = resumen_limpio.copy()
resumen_limpio_v['version'] = 'sin duplicados'

# pd.concat() apila los dos resumenes uno debajo del otro (axis=0 por defecto).
# Ordenamos por 'hotel' para que las dos versiones de cada hotel queden juntas y
# renumeramos el indice con reset_index(drop=True) para que quede consecutivo.
comparacion = pd.concat([resumen_original_v, resumen_limpio_v], ignore_index=True)
comparacion = comparacion.sort_values(by='hotel').reset_index(drop=True)

comparacion


,hotel,n_reservas,tasa_cancelacion,adr_promedio,version
0,City Hotel,79330,0.417270,105.304465,original
1,City Hotel,53428,0.300386,110.985944,sin duplicados
2,Resort Hotel,40060,0.277634,94.952930,original
3,Resort Hotel,33968,0.234809,99.025346,sin duplicados


### 3.4 Interpretación (6 pts)

a) ¿Qué métrica cambia más al eliminar duplicados? ¿En qué hotel el cambio es mayor? Explica por qué, relacionándolo con lo encontrado en la Actividad 2.

b) Si se entrenara un modelo para predecir `is_canceled` con los datos originales, ¿qué problemas causarían los duplicados en el entrenamiento y en la evaluación del modelo?

**Respuesta 3.4:**

**a)** La metrica que **mas cambia** al eliminar duplicados es la **tasa de cancelacion**. En el City Hotel pasa de 41.73% (original) a 30.04% (sin duplicados), una caida de casi 12 puntos porcentuales; en el Resort Hotel pasa de 27.76% a 23.48%, una caida de solo 4.3 puntos. El **cambio es mayor en el City Hotel**, tanto en puntos porcentuales como en numero de reservas afectadas (pierde 25,902 filas contra 6,092 del Resort Hotel).

Esto es consistente con lo encontrado en la Actividad 2: las filas duplicadas tienen una tasa de cancelacion mucho mas alta que el promedio general (58.39% contra 37.04%), y el City Hotel concentra la mayor proporcion de duplicados (32.65% de sus reservas, frente a 15.21% del Resort Hotel). Al eliminar esas copias "infladas" de reservas canceladas, la tasa de cancelacion del City Hotel baja mucho mas que la del Resort Hotel. El `adr` promedio tambien sube ligeramente en ambos hoteles al quitar duplicados (de 105.30 a 110.99 en City Hotel, de 94.95 a 99.03 en Resort Hotel), aunque el cambio es menor que en la tasa de cancelacion.

**b)** Entrenar un modelo de `is_canceled` con los datos originales tendria varios problemas:

- **Fuga entre entrenamiento y prueba (data leakage):** si el conjunto de entrenamiento y el de prueba se separan al azar, es muy probable que copias identicas de una misma reserva caigan una en cada conjunto. El modelo "veria" en entrenamiento filas identicas a las que luego debe predecir en prueba, lo que infla artificialmente la precision evaluada sin que el modelo realmente generalice.
- **Sesgo en el aprendizaje:** como las reservas canceladas estan sobre-representadas entre los duplicados (58.39% contra 37.04% general), el modelo aprenderia con un peso desproporcionado hacia ciertos patrones de reservas grupales o con deposito no reembolsable, sesgando sus predicciones y sobreestimando la probabilidad de cancelacion para ese tipo de reservas en la poblacion general.
- **Metricas de evaluacion poco confiables:** el desempeno reportado (exactitud, AUC, etc.) no reflejaria el desempeno real del modelo sobre reservas nuevas, sino un desempeno inflado por la memorizacion de filas repetidas.


---
## Actividad 4. Tratamiento y verificación (20 puntos)

### 4.1 Verificación (5 pts)

Verifica sobre `hoteles_limpio` que:

- No quedan duplicados exactos.
- No hay índices repetidos y el índice va de 0 a `len(hoteles_limpio) - 1`.

In [22]:
# Verificamos que ya no existan filas duplicadas exactas en hoteles_limpio.
duplicados_restantes = hoteles_limpio.duplicated().sum()
print(f"Duplicados exactos restantes: {duplicados_restantes}")

# Verificamos que no haya indices repetidos.
indices_repetidos = hoteles_limpio.index.duplicated().sum()
print(f"Indices repetidos: {indices_repetidos}")

# Verificamos que el indice sea exactamente el rango de 0 a len(hoteles_limpio) - 1.
indice_correcto = (hoteles_limpio.index == range(len(hoteles_limpio))).all()
print(f"El indice va de 0 a len(hoteles_limpio)-1 sin huecos: {indice_correcto}")


Duplicados exactos restantes: 0
Indices repetidos: 0
El indice va de 0 a len(hoteles_limpio)-1 sin huecos: True


### 4.2 Duplicados parciales (10 pts)

La columna `reservation_status_date` registra la fecha del último cambio de estado. Detecta las filas de `hoteles_limpio` que **coinciden en todas las columnas excepto `reservation_status_date`**.

1. Cuenta cuántas filas sobrantes hay con ese criterio.
2. Muestra algunas de esas filas ordenadas, con las columnas `hotel`, `lead_time`, `arrival_date_year`, `arrival_date_month`, `arrival_date_day_of_month`, `adr`, `reservation_status` y `reservation_status_date`.
3. Muestra la distribución de `reservation_status` en esas filas.

Después responde: ¿qué podrían representar estas filas? ¿Las eliminarías? Justifica.

In [23]:
# Definimos el conjunto de columnas a comparar: todas EXCEPTO 'reservation_status_date'.
columnas_sin_fecha_estado = [c for c in hoteles_limpio.columns if c != 'reservation_status_date']

# 1. Filas sobrantes bajo este criterio (manteniendo la primera aparicion de cada combinacion).
duplicados_parciales_sobrantes = hoteles_limpio.duplicated(subset=columnas_sin_fecha_estado, keep='first').sum()
print(f"Filas sobrantes que coinciden en todo excepto en reservation_status_date: {duplicados_parciales_sobrantes}")

# Tambien identificamos TODAS las filas involucradas (keep=False) para poder inspeccionarlas
# y ver su distribucion de reservation_status.
mascara_duplicados_parciales = hoteles_limpio.duplicated(subset=columnas_sin_fecha_estado, keep=False)
filas_dup_parciales = hoteles_limpio[mascara_duplicados_parciales]
print(f"Filas involucradas en estos grupos (keep=False): {mascara_duplicados_parciales.sum()}")

# 2. Mostramos una muestra ordenada para poder comparar visualmente las filas de un mismo grupo.
columnas_vista = ['hotel', 'lead_time', 'arrival_date_year', 'arrival_date_month',
                   'arrival_date_day_of_month', 'adr', 'reservation_status', 'reservation_status_date']

muestra_ordenada = filas_dup_parciales.sort_values(by=columnas_sin_fecha_estado)
print("\nMuestra de filas duplicadas parciales (difieren solo en la fecha de estado):")
print(muestra_ordenada[columnas_vista].head(10))

# 3. Distribucion de reservation_status en estas filas.
print("\nDistribucion de reservation_status en las filas con duplicado parcial:")
print(filas_dup_parciales['reservation_status'].value_counts())


Filas sobrantes que coinciden en todo excepto en reservation_status_date: 245
Filas involucradas en estos grupos (keep=False): 471

Muestra de filas duplicadas parciales (difieren solo en la fecha de estado):
            hotel  lead_time  arrival_date_year arrival_date_month  \
35572  City Hotel          9               2015          September   
35574  City Hotel          9               2015          September   
51267  City Hotel         12               2017               July   
51268  City Hotel         12               2017               July   
54170  City Hotel         14               2015            October   
54474  City Hotel         14               2015            October   
39425  City Hotel         14               2016              April   
39432  City Hotel         14               2016              April   
40080  City Hotel         14               2016                May   
40083  City Hotel         14               2016                May   

       arrival_date_

**Respuesta 4.2:**

Se encontraron **471 filas** (agrupadas en 226 grupos, es decir, 245 filas "sobrantes" con `keep='first'`) que coinciden en absolutamente todas las columnas excepto en `reservation_status_date`. Lo mas revelador es que el **100% de esas filas tiene `reservation_status = 'Canceled'`**.

Esto sugiere que estas filas probablemente representan **el mismo registro de reserva capturado o actualizado en dos momentos distintos del proceso de cancelacion** (por ejemplo, una primera fecha en la que se marco como cancelada y una fecha posterior de confirmacion o de sincronizacion del sistema), y no reservas genuinamente diferentes: es poco creible que dos clientes distintos coincidan en el hotel, la fecha de llegada, el numero de huespedes, el `adr` exacto y todas las demas variables, y que ademas ambos hayan cancelado.

**¿Las eliminaria?** Si, tratandolas como duplicados y conservando (por ejemplo) la fila con la fecha de estado mas reciente, ya que representa el estado final/mas actualizado de esa reserva. Antes de decidir de forma definitiva en un proyecto real, valdria la pena confirmar con el proveedor de los datos si `reservation_status_date` puede legitimamente registrar mas de un evento por reserva (lo cual justificaria conservar ambas filas como un historial) o si, como parece mas probable aqui, se trata de una duplicacion en el proceso de carga o actualizacion de los datos.


### 4.3 Registro de decisión (5 pts)

Redacta, en una celda de texto, el registro de la limpieza realizada, como lo documentarías en un proyecto real. Incluye:

- **Criterio** de duplicado utilizado (columnas comparadas).
- **Regla de conservación** (`keep`) y por qué es o no relevante en este caso.
- **Resultado** (filas antes, después y eliminadas).
- **Supuestos y limitaciones**.

**Respuesta 4.3:**

**Registro de decisión de limpieza — Dataset Hotel Booking Demand**

- **Criterio de duplicado utilizado:** se consideraron duplicadas las filas que coinciden en las 32 columnas del dataset (duplicado exacto, sin excluir ninguna variable). Este es el criterio mas estricto posible y el mas conservador, ya que solo elimina filas cuando existe una coincidencia total.

- **Regla de conservacion (`keep`):** se utilizo `keep='first'`, es decir, se conservo la primera aparicion de cada combinacion de valores y se eliminaron las copias posteriores. En este caso la eleccion de `keep` **no es relevante para las metricas agregadas** (conteos, medias, tasas), porque al ser copias identicas, conservar la primera o la ultima produce exactamente los mismos resultados numericos; solo cambiaria si se necesitara preservar informacion asociada al orden original de carga de los datos (por ejemplo, si la posicion en el archivo reflejara el orden cronologico real de registro).

- **Resultado:** el dataset paso de 119,390 filas a 87,396 filas, es decir, se eliminaron 31,994 filas duplicadas (26.80% del total). Se verifico que `hoteles_limpio` no contiene duplicados exactos remanentes y que su indice es consecutivo de 0 a 87,395 sin huecos ni repeticiones.

- **Supuestos y limitaciones:**
  1. Se asume que una coincidencia exacta en las 32 columnas (incluyendo variables continuas como `adr`) es extremadamente improbable entre dos reservas genuinamente distintas, por lo que se interpretan como duplicados de registro y no como coincidencias reales.
  2. `duplicated()` trata los valores faltantes (`NaN`) en `agent` y `company` como iguales entre si, lo que pudo haber sobreestimado ligeramente el numero de duplicados en reservas sin agencia/empresa asignada.
  3. Al no existir un identificador unico de reserva en el dataset, no es posible confirmar con certeza absoluta el origen de los duplicados (error de carga, reintento del sistema, reserva re-enviada, etc.); la decision se basa en evidencia indirecta (tamano de los grupos, concentracion en ciertos segmentos, tasa de cancelacion).
  4. Se detectaron ademas 471 filas de "duplicado parcial" (identicas salvo en `reservation_status_date`), que no se eliminaron en este paso por tratarse de un criterio distinto al de duplicado exacto solicitado en la Actividad 3, pero que se documentan como un hallazgo adicional a revisar (ver Actividad 4.2).


---
## Reflexión final (10 puntos)

Responde brevemente, con base en tus resultados:

1. Describe una situación en la que **no** convendría eliminar los duplicados exactos de este dataset.
2. ¿Qué información adicional (columnas) le pedirías al proveedor de los datos para tomar una decisión más segura? ¿Por qué?
3. En tus propias palabras, ¿por qué `duplicated().sum() == 0` no garantiza que un dataset esté libre de duplicados?

**Respuesta final:**

1. **Situacion en la que no convendria eliminar duplicados exactos:** si el objetivo del analisis fuera **medir el volumen de solicitudes o transacciones procesadas por el sistema de reservas** (por ejemplo, para dimensionar la carga del servidor o auditar cuantas veces se reenvio una reserva al sistema), entonces cada fila —aunque sea identica a otra— representaria un evento real de procesamiento, y eliminarla haria perder informacion operativa relevante, incluso si para el analisis de negocio (demanda real de habitaciones) esas mismas filas si deban tratarse como duplicados.

2. **Informacion adicional que pediria al proveedor:** principalmente un **identificador unico de reserva** (`booking_id`) y, si es posible, un **timestamp de creacion/actualizacion del registro** (no solo la fecha de `reservation_status`). Con un identificador de reserva se podria distinguir de forma definitiva si dos filas identicas son la misma reserva registrada dos veces o coincidencias genuinas entre reservas distintas, eliminando la necesidad de inferirlo indirectamente por el tamano de los grupos o la distribucion de variables como se hizo en la Actividad 2.

3. **Por que `duplicated().sum() == 0` no garantiza un dataset libre de duplicados:** `duplicated()` solo detecta duplicados **exactos** en las columnas que se le indiquen (por defecto, todas). Un dataset puede no tener ninguna fila identica en el sentido estricto y, sin embargo, seguir teniendo **duplicados logicos o parciales**: por ejemplo, la misma reserva registrada con una pequena diferencia en una sola columna (como se vio en la Actividad 4.2 con `reservation_status_date`), errores de captura (mayusculas/minusculas distintas, espacios extra), o registros que representan el mismo evento del mundo real bajo una codificacion ligeramente distinta. Por eso, un `duplicated().sum() == 0` solo certifica ausencia de duplicados **exactos en las columnas evaluadas**, no ausencia de duplicados en un sentido mas amplio.


---
**Fin de la actividad.** Antes de entregar: **Reiniciar y ejecutar todo**, verificar que no haya errores, guardar con las salidas visibles y comprobar que el enlace abre sin iniciar sesión.